relayer/message_coordinator.go

```diff
@@ -0,0 +1,254 @@
+	applicationRelayerID = database.CalculateRelayerID(
+		sourceBlockchainID,
+		destinationBlockchainID,
+		originSenderAddress,
+		database.AllAllowedAddress,
+	)
+	if applicationRelayer, ok := mc.applicationRelayers[applicationRelayerID]; ok {
+		return applicationRelayer
+	}
+
+	// Check for a match on sourceBlockchainID and destinationBlockchainID, with any originSenderAddress and a specific destinationAddress
+	applicationRelayerID = database.CalculateRelayerID(
+		sourceBlockchainID,
+		destinationBlockchainID,
+		database.AllAllowedAddress,
+		destinationAddress,
+	)
+	if applicationRelayer, ok := mc.applicationRelayers[applicationRelayerID]; ok {
+		return applicationRelayer
+	}
+
+	// Check for a match on sourceBlockchainID and destinationBlockchainID, with any originSenderAddress and any destinationAddress
+	applicationRelayerID = database.CalculateRelayerID(
+		sourceBlockchainID,
+		destinationBlockchainID,
+		database.AllAllowedAddress,
+		database.AllAllowedAddress,
+	)
+	if applicationRelayer, ok := mc.applicationRelayers[applicationRelayerID]; ok {
+		return applicationRelayer
+	}
+	mc.logger.Debug(
+		"Application relayer not found. Skipping message relay.",
+		zap.String("blockchainID", sourceBlockchainID.String()),
+		zap.String("destinationBlockchainID", destinationBlockchainID.String()),
+		zap.String("originSenderAddress", originSenderAddress.String()),
+		zap.String("destinationAddress", destinationAddress.String()),
+	)
+	return nil
+}
+
+func (mc *MessageCoordinator) ProcessManualWarpMessage(
+	warpMessage *relayerTypes.WarpMessageInfo,
+) (common.Hash, error) {
+	// Send any messages that were specified in the configuration
+	mc.logger.Info(
+		"Relaying manual Warp message",
+		zap.String("blockchainID", warpMessage.UnsignedMessage.SourceChainID.String()),
+		zap.String("warpMessageID", warpMessage.UnsignedMessage.ID().String()),
+	)
+	appRelayer, handler, err := mc.GetAppRelayerMessageHandler(warpMessage)
+	if err != nil {
+		mc.logger.Error(
+			"Failed to parse manual Warp message.",
+			zap.Error(err),
+			zap.String("warpMessageID", warpMessage.UnsignedMessage.ID().String()),
+		)
+		return common.Hash{}, err
+	}
+
+	return appRelayer.ProcessMessage(handler)
+}
+
+func (mc *MessageCoordinator) ProcessMessage(blockchainID ids.ID, messageID common.Hash, blockNum *big.Int) (common.Hash, error) {
+	ethClient, ok := mc.sourceClients[blockchainID]
+	if !ok {
+		mc.logger.Error("Source client not found", zap.String("blockchainID", blockchainID.String()))
+		return common.Hash{}, fmt.Errorf("source client not set for blockchain: %s", blockchainID.String())
+	}
+
+	warpMessage, err := relayerTypes.FetchWarpMessageFromID(ethClient, messageID, blockNum)
+	if err != nil {
+		mc.logger.Error("Failed to fetch warp from blockchain", zap.String("blockchainID", blockchainID.String()), zap.Error(err))
+		return common.Hash{}, fmt.Errorf("could not fetch warp message from ID: %w", err)
+	}
```

Comment:
To simplify the `MessageCoorinator` exported interface, how about we lift this logic out to the caller, and combine `ProcessMessage` and `ProcessManualWarpMessage`?

Reply:
Good call :)

Reply:
Yeah, we'd need to decide if the interface should accept only Warp Messages, or also IDs. Only accepting Warp Messages would yield a smaller interface, but would require more work on the caller's part as you described. I'm good to leave this as is for now.